designs/2019-expose-rules-to-formatters/readme.MD

```diff
@@ -0,0 +1,137 @@
+- Start Date: 2019-01-16
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: Chris Meyer (@EasyRhinoMSFT)
+
+# Providing Rule Metadata to Formatters
+
+## Summary
+
+This proposal describes a design enhancement that provides formatters with details about the rules that have been executed by ESLint.
+
+## Motivation
+
+Currently, formatters only see the ID of each rule for which a violation was identified, plus an instance-specific description, as properties on each result object. Formatters are not able to access usefule rule metadata, such as category, description, and help URL. Formatters are also not aware of the full set of rules that were run, information that may be useful in some cases.
+
+## Detailed Design
+
+Design Summary
+1. In `cli.js::printResults`, obtain the rules map from the `Engine` object.
+2. Pass the rules map as property in a new, second parameter object to the formatter's exported interface function.
```

Comment:
The motivation section indicated that the primary use case of this RFC is to allow formatters to access rule metadata. It's not clear to me that formatters need to have access to rule objects themselves, so I would propose only giving them access to metadata. (In general, I think it's better to avoid providing unnecessary information in an API, since it can make the implementation more difficult to change later.)

In this case, I think the formatter invocation would look like this:

```js
const output = formatter(
    results,
    {
        ruleMetadata: {
            "accessor-pairs": rules.get("accessor-pairs").meta,
            "array-bracket-newline": rules.get("array-bracket-newline").meta,
            // ...
        }
    }
);
```

(The object would be dynamically generated rather than hardcoded; I just hardcoded rules here to make the explanation easier to understand.)